common/exception/src/exception.rs

```diff
@@ -257,3 +257,50 @@ impl ErrorCodes {
         }
     }
 }
+
+/// Provides the `map_err_to_code` method for `Result`.
+///
+/// ```
+/// use common_exception::ToErrorCodes;
+/// use common_exception::ErrorCodes;
+///
+/// let x: std::result::Result<(), std::fmt::Error> = Err(std::fmt::Error {});
+/// let y: common_exception::Result<()> =
+///     x.map_err_to_code(ErrorCodes::UnknownException, || format!("{}", 123));
+///
+/// assert_eq!(
+///     "Code: 1000, displayText = Error.",
+///     format!("{}", y.unwrap_err())
+/// );
+/// ```
+pub trait ToErrorCodes<T, E, ToStrFn> {
+    /// Wrap the error value with ErrorCodes that is evaluated lazily
+    /// only once an error does occur.
+    ///
+    /// `err_code_fn` is one of the ErrorCodes builder function such as `ErrorCodes::Ok`.
+    /// `context_fn` builds display_text for the ErrorCodes.
+    fn map_err_to_code<ToStr, ErrFn>(self, err_code_fn: ErrFn, context_fn: ToStrFn) -> Result<T>
+    where
+        ToStr: Into<String>,
+        ErrFn: FnOnce(ToStr) -> ErrorCodes,
+        ToStrFn: FnOnce() -> ToStr;
+}
+
+impl<T, E, ToStrFn> ToErrorCodes<T, E, ToStrFn> for std::result::Result<T, E>
+where E: std::error::Error + Send + Sync + 'static
+{
+    fn map_err_to_code<ToStr, ErrFn>(self, make_exception: ErrFn, context_fn: ToStrFn) -> Result<T>
+    where
+        ToStr: Into<String>,
+        ErrFn: FnOnce(ToStr) -> ErrorCodes,
+        ToStrFn: FnOnce() -> ToStr,
+    {
+        self.map_err(|error| {
+            let mut x = make_exception(context_fn());
+            x.cause = Some(Box::new(error));
```

Comment:
Using 'cause' may not be a good idea. for example, we can't transfer them in multiple nodes. maybe `make_exception(format!("{}, cause : {}", context_fn(), error))` is better?

Reply:
maybe we can remove `cause`?

Reply:
It reminds me of the error message I saw a few days before:

![2021-06-03-164456_1917x219_scrot](https://user-images.githubusercontent.com/22081156/120619880-fdf4a180-c48e-11eb-99c5-cfc055c8a3c0.png)


would the `"message":"Unknown database: 'db1'"` part  be generated by  `cause` (if cause is `Some<Box<..>>`)?

Reply:
How can be reproduce(this is because errors are nested)? I think we should fix it.

Reply:
> It reminds me of the error message I saw a few days before:
> 
> ![2021-06-03-164456_1917x219_scrot](https://user-images.githubusercontent.com/22081156/120619880-fdf4a180-c48e-11eb-99c5-cfc055c8a3c0.png)
> 
> could the `"message":"Unknown database: 'db1'"` part be generated based on `ErrorCodes::cause` (if cause is `Some<Box<..>>`) by method `ErrorCodes::message`?

It seems like this redundant message is caused by using `Debug::fmt()` on `ErrorCodes::cause`:

https://github.com/datafuselabs/datafuse/blob/668ef027b6f4bb713115edc2c7b1db7f30c9184e/common/exception/src/exception.rs#L45-L50

The `message()` should format `cause` with `Display` instead of `Debug`. 
As a `Display` impl for `std::error::Error` outputs the description of the error:
```
    fn description(&self) -> &str {
        "description() is deprecated; use Display"
    }

```

Reply:
> maybe we can remove `cause`? it's only used in https://github.com/datafuselabs/datafuse/blob/master/common/exception/src/exception.rs#L195

`std::error::Error` has a similar term: `cause()`. To be consistent, I'd like to have a `cause` in `ErrorCodes`. :D

Reply:
> Using 'cause' may not be a good idea. for example, we can't transfer them in multiple nodes. maybe `make_exception(format!("{}, cause : {}", context_fn(), error))` is better?
> 
> https://github.com/datafuselabs/datafuse/blob/master/fusequery/query/src/api/rpc/mod.rs#L32-L68

🤔 I do not see a problem here(Remind me if I missed anything): we can just recursively serialize an `ErrorCdoes` and its `cause`. I.e. chaining `SerializedError` with `cause` too: 
```
struct SerializedError {
    code: u16,
    message: String,
    cause: Box<SerializedError>,
}
```

Is it possible?

Reply:
the "unknown database" error no longer exists, it happens in my local buggy branch  ;-(  already fixed.
 
 I forget to mention that the "nested error" `displayText = {"code":3,"....` ; was reformated, `\n` characters are replaced by newline.

Reply:
> 🤔 I do not see a problem here(Remind me if I missed anything): we can just recursively serialize an ErrorCdoes and its cause. I.e. chaining SerializedError with cause too:
>
> struct SerializedError {
>     code: u16,
>     message: String,
>     cause: Box\<SerializedError\>,
> }
> Is it possible?

When cause is not ErrorCodes. this breaks the recursion. 
When cause is ErrorCodes, can this is useless?

Reply:
> When cause is ErrorCodes, can this is useless?

Definitely not, bro. It helps me a lot while debugging the multi-node cases. 

Maybe we could discuss it at the meetup tomorrow?  @zhang2014  @drmingdrmer

Reply:
ok, meetup#topoc1

Reply:
> When cause is not ErrorCodes. this breaks the recursion.

If `cause` is not ErrorCdoes:
- just stop the recursion and append `cause.description()` to `message`,
- or build a special `SerializedError` without a valid `code` to continue the recursion. Since the std error has a method `cause()` too.